python/mujinwebstackclient/webstackgraphclientutils.py

```diff
@@ -89,3 +92,197 @@ def _CallSimpleGraphAPI(self, queryOrMutation, operationName, parameterNameTypeV
         if log.isEnabledFor(5): # logging.VERBOSE might not be available in the system
             log.verbose('got response from graph query: %r', data)
         return data.get(operationName)
+
+def UseGraphQueryResult(queryFunction):
+    """This decorator break a large graph query into a few small queries with the help of QueryResult class to prevent webstack from consuming too much memory.
+    """
+    @wraps(queryFunction)
+    def inner(self, *args, **kwargs):
+        queryResult = GraphQueryResult(queryFunction, *((self,) + args), **kwargs)
+        response = {}
+        if queryResult.keyName == '__typename':
+            response['__typename'] = queryResult[0]
+        elif queryResult.keyName is not None:
+            response[queryResult.keyName] = queryResult
+        if queryResult.totalCount is not None:
+            response['meta'] = {'totalCount': queryResult.totalCount}
+        return response
+
+    return inner
+
+class GraphQueryResult(webstackclientutil.QueryResult):
+    """Wraps graph query response. Break large query into small queries automatically to save memory.
+    """
+    _totalCount = None
+    _keyName = None
+
+    def __init__(self, queryFunction, *args, **kwargs):
+        self._queryFunction = queryFunction
+        self._args = args
+        self._kwargs = copy.deepcopy(kwargs)
+        if self._kwargs.get('options', None) is None:
+            self._kwargs['options'] = {'offset': 0, 'first': 0}
+        self._kwargs['options'].setdefault('offset', 0)
+        self._kwargs['options'].setdefault('first', 0)
+        self._limit = self._kwargs['options']['first']
+        self._offset = self._kwargs['options']['offset']
+        if 'fields' in self._kwargs:
+            self._kwargs.setdefault('fields', {})
+            self._kwargs['fields'].setdefault('meta', {})
+            self._kwargs['fields']['meta'].setdefault('totalCount', None)
+        self._hasCompleteQueryResult = False
+        self._APICall(offset=self._offset)
+
+    def __iter__(self):
+        if self._hasCompleteQueryResult:
+            return list.__iter__(self)
+        return GraphQueryIterator(self._queryFunction, *self._args, **self._kwargs)
+    
+    def _APICall(self, offset):
```

Comment:
`_CallAPI` to follow the styleguide

Reply:
We already have a function `APICall` in `controllerwebclientraw.py`. I think reusing the same name here is better. Or maybe `_QueryWebstack` to distinguish them?

Reply:
I see